Inject transactions from past eras when forging a block

mrBliss opened the pull request:
When the mempool verifies transactions, it injects transactions from past eras
into the current era. However, block production was not doing this.

Debugging this was made more difficult because the trace message that we were
outputting after creating a block (`TraceAdoptedBlock`) recorded the
transactions from the mempool, not the transactions that were actually included
in the block.

This means that injection now happens 3 times: once when we add transactions
into the mempool, once when we revalidate the mempool when we produce a block,
and then a third time when we actually construct the block. Streamlining this
however we can do at a later stage.

edsko reviewed (approved):
Been a while you and I had a debugging session like this :grimacing:

intricate commented on ouroboros-consensus/src/Ouroboros/Consensus/HardFork/Combinator/Forging.hs:
```diff
@@ -298,45 +302,69 @@ hardForkForgeBlock blockForging
+             ([Match.Mismatch GenTx f xs], NS (Product f ([] :.: GenTx)) xs)
+           -> NS (Product f ([] :.: GenTx)) xs
+        noMismatches ([], xs)  = xs
+        noMismatches (errs, _) = error "unexpected unmatchable transactions"
```
```suggestion
        noMismatches (_errs, _) = error "unexpected unmatchable transactions"
```

dcoutts reviewed (approved):
Very nice.

I added the smallest patch I could think of to make this build. It involves reverting the addition of a sensible assertion so that an extra `HasTxs` constraint is not required.

@mrBliss Feel free to re-enable the assertion if you can resolve the `HasTxs` issue.

dcoutts commented:
Oh and feel free to squash these little fixup patches too of course.

dcoutts commented:
One test failure.

Looks unrelated to this PR, but perhaps is related to the recent patch to support wall-clock times moving backwards by a small amount?
```
    delayClockShift:                         FAIL (0.31s)
      *** Failed! Falsified (after 78 tests and 5 shrinks):
      Schedule {getSchedule = [11.5,9.5]}
      Expected exception but got [SlotNo 0,SlotNo 11,SlotNo 12,SlotNo 13,SlotNo 14,SlotNo 15,SlotNo 16,SlotNo 17,SlotNo 18,SlotNo 19,SlotNo 20,SlotNo 21,SlotNo 22,SlotNo 23,SlotNo 24,SlotNo 25,SlotNo 26,SlotNo 27,SlotNo 28,SlotNo 29,SlotNo 30,SlotNo 31,SlotNo 32,SlotNo 33,SlotNo 34,SlotNo 35,SlotNo 36,SlotNo 37,SlotNo 38,SlotNo 39,SlotNo 40,SlotNo 41,SlotNo 42,SlotNo 43,SlotNo 44,SlotNo 45,SlotNo 46,SlotNo 47,SlotNo 48,SlotNo 49,SlotNo 50,SlotNo 51,SlotNo 52,SlotNo 53,SlotNo 54,SlotNo 55,SlotNo 56,SlotNo 57,SlotNo 58,SlotNo 59,SlotNo 60,SlotNo 61,SlotNo 62,SlotNo 63,SlotNo 64,SlotNo 65,SlotNo 66,SlotNo 67,SlotNo 68,SlotNo 69,SlotNo 70,SlotNo 71,SlotNo 72,SlotNo 73,SlotNo 74,SlotNo 75,SlotNo 76,SlotNo 77,SlotNo 78,SlotNo 79,SlotNo 80,SlotNo 81,SlotNo 82,SlotNo 83,SlotNo 84,SlotNo 85,SlotNo 86,SlotNo 87,SlotNo 88,SlotNo 89,SlotNo 90,SlotNo 91,SlotNo 92,SlotNo 93,SlotNo 94,SlotNo 95,SlotNo 96,SlotNo 97,SlotNo 98,SlotNo 99,SlotNo 100,SlotNo 101,SlotNo 102,SlotNo 103,SlotNo 104,SlotNo 105,SlotNo 106,SlotNo 107,SlotNo 108,SlotNo 109]
```

mrBliss commented:
Squashed and fixed the test failure. Unfortunately, reintroducing the assertion based on `HasTxs` is not trivial (a known problem we forgot about), so I'm leaving that out for now.

mrBliss commented:
bors merge
